### PR TITLE
adds new edit flow to the media+text block

### LINKED ISSUE #14795: Expand the new replace image flow out to other blocks
#14142 added an improved flow for replacing an image: 

swap

As per @mapk's [note in the thread](https://github.com/WordPress/gutenberg/pull/14142#issuecomment-476871253), this flow should be rolled out to other areas of the UI. Here's a quick list: 

- [ ] Video
- [ ] Audio
- [ ] Embeds* (Fixing this will also fix #11439)

In the following blocks, the edit button in the toolbar opens the media modal directly, so they just need the pencil icon swapped out for the new "Replace" icon: 

- [ ] Media & Text 
- [ ] Cover
- [ ] File

---

_*In the case of the Embed blocks (and maybe Video + Audio?) showing the preview in the placeholder may be too complicated. If that's the case, it may be fine to leave that part out._

### PR DESCRIPTION
**Pull request by draganescu**
# Description

Closes #14795

This is a follow up to the image block edit flow update which ports the new flow to all the blocks which have media with an edit state.

# How has this been tested?
For now I've only tested locally.

# Types of change
New feature (non-breaking change which adds functionality)

# Screenshots

media+text

### CODE REVIEW

**Review by kjellr — COMMENTED** [review]
This is working well for me. The only small visual issue I'm seeing is the issue @talldan noted in [the original thread](https://github.com/WordPress/gutenberg/pull/14918#issuecomment-488616390): 

> Some space is needed between the Media Library and Insert From URL buttons (seems to also apply to other blocks). This does seem to be an issue with the image block as well, so could always fix it after this PR for all blocks:
> 

I agree we could fix this separately if it's not a super-quick global fix.

File: packages/block-library/src/media-text/edit.js [hunk 1]
```diff
@@ -102,14 +141,25 @@ class MediaTextEdit extends Component {
...
+		const {
+			mediaAlt,
+			mediaId,
+			mediaPosition,
+			mediaType,
+			mediaUrl,
+			mediaWidth,
+			imageFill,
+			focalPoint,
+		} = attributes;
```
Comment (talldan) [inline packages/block-library/src/media-text/edit.js]: some line breaks between different chunks of code would be nice to make this more readable.

File: packages/block-library/src/media-text/edit.js [hunk 2]
```diff
@@ -39,17 +44,47 @@ const TEMPLATE = [
...
+
+		if ( newURL !== mediaUrl ) {
+			this.props.setAttributes( {
+				mediaUrl: newURL,
+				mediaId: undefined,
+			} );
+		}
+
+		this.setState( {
+			isEditing: false,
```
Comment (talldan) [inline packages/block-library/src/media-text/edit.js]: Should this trigger `speak( __( 'You are now viewing the image in the image block.' ) )`?

File: packages/block-library/src/media-text/edit.js [hunk 3]
```diff
@@ -73,6 +108,10 @@ class MediaTextEdit extends Component {
 			src = get( media, [ 'sizes', 'large', 'url' ] ) || get( media, [ 'media_details', 'sizes', 'large', 'source_url' ] );
 		}
 
+		this.setState( {
+			isEditing: false,
```
Comment (talldan) [inline packages/block-library/src/media-text/edit.js]: Should this trigger `speak( __( 'You are now viewing the image in the image block.' ) )`?

File: packages/block-library/src/media-text/edit.js [hunk 4]
```diff
@@ -39,17 +44,47 @@ const TEMPLATE = [
...
+		}
+	}
+
+	onSelectUrl( newURL ) {
+		const { mediaUrl } = this.props.attributes;
+
+		if ( newURL !== mediaUrl ) {
+			this.props.setAttributes( {
+				mediaUrl: newURL,
+				mediaId: undefined,
```
Comment (talldan) [inline packages/block-library/src/media-text/edit.js]: Are there any other media properties that should be unset when replacing media?

File: packages/block-library/src/media-text/editor.scss [hunk 5]
```diff
@@ -50,6 +50,10 @@ figure.block-library-media-text__media-container {
 	width: 100%;
 }
 
+.wp-block-media-text .block-library-media-text__media-container .components-placeholder__preview img {
```
Comment (talldan) [inline packages/block-library/src/media-text/editor.scss]: This is a pretty long selector, were there specificity issues? If so, it'd be great to add a comment mentioning it.

File: packages/block-library/src/media-text/media-container.js [hunk 6]
```diff
@@ -80,60 +53,84 @@ class MediaContainer extends Component {
...
-			};
-			const onResizeStop = ( event, direction, elt ) => {
-				commitWidthChange( parseInt( elt.style.width ) );
-			};
-			const enablePositions = {
-				right: mediaPosition === 'left',
-				left: mediaPosition === 'right',
-			};
+		const { isEditing, mediaPosition, mediaType, mediaWidth, commitWidthChange, onWidthChange } = this.props;
+		if ( isEditing ) {
```
Comment (talldan) [inline packages/block-library/src/media-text/media-container.js]: Being able to unwrap that if statement is a nice improvement to this method.

**Comment by draganescu** [comment]
To have proper spacing between buttons we need a better structure in media-placeholder output b/c doing it via CSS adds a lot of bloat for no reason. Will open a separate PR for that :) @talldan @kjellr

File: packages/block-library/src/media-text/edit.js (hunk 4 above)
Comment (draganescu) [reply packages/block-library/src/media-text/edit.js]: yes, found some more being set by onSelectMedia.

File: packages/block-library/src/media-text/media-container.js (hunk 6 above)
Comment (draganescu) [reply packages/block-library/src/media-text/media-container.js]: thanks!

**Comment by kjellr** [comment]
> To have proper spacing between buttons we need a better structure in media-placeholder output b/c doing it via CSS adds a lot of bloat for no reason. Will open a separate PR for that :)

Sounds good. This PR is looking pretty good in that case. The only bug I'm seeing now is that while I can make use of the "Insert from URL" option as when logged in as a Contributor, the cancel link doesn't display under the button: 

Screen Shot 2019-05-27 at 11 32 13 AM

**Comment by mapk** [comment]
I tested this today and it works well. I ran through a few configurations and it was great! Thanks for your work on this @draganescu 💯

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @draganescu the comment I left on the cover block https://github.com/WordPress/gutenberg/pull/15519#pullrequestreview-245946719  also applied here. I think we should have a solution used in both blocks.

**Comment by draganescu** [comment]
#11952 changed direction so closing this as it became irrelevant.

**Pull request closed without merging**